nav2_collision_monitor/src/polygon_base.cpp

```diff
@@ -0,0 +1,175 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/polygon_base.hpp"
+#include "nav2_collision_monitor/kinematics.hpp"
+
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_collision_monitor
+{
+
+PolygonBase::PolygonBase()
+: polygon_type_(POLYGON_BASE), polygon_name_(""), action_type_(DO_NOTHING),
+  stop_points_(-1), slowdown_(0.0), time_before_collision_(-1.0),
+  simulation_time_step_(0.0)
+{
+}
+
+PolygonBase::PolygonBase(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const std::string polygon_name,
+  const double simulation_time_step)
+: node_(node), polygon_type_(POLYGON_BASE), polygon_name_(polygon_name), action_type_(DO_NOTHING),
+  stop_points_(-1), slowdown_(0.0), time_before_collision_(-1.0),
+  simulation_time_step_(simulation_time_step)
+{
+  auto node_sptr = node_.lock();
+  if (node_sptr) {
+    RCLCPP_INFO(node_sptr->get_logger(), "Creating PolygonBase");
+  }
+}
+
+PolygonBase::~PolygonBase()
+{
+  auto node = node_.lock();
+  if (node) {
+    RCLCPP_INFO(node->get_logger(), "Destroying PolygonBase");
+  }
+}
+
+bool PolygonBase::getParameters()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  try {
+    // Get action type
+    nav2_util::declare_parameter_if_not_declared(
+      node, polygon_name_ + ".action_type", rclcpp::ParameterValue("stop"));  // Stop by default
+    const std::string at_str =
+      node->get_parameter(polygon_name_ + ".action_type").as_string();
+    if (at_str == "stop") {
+      action_type_ = STOP;
+    } else if (at_str == "slowdown") {
+      action_type_ = SLOWDOWN;
+    } else {  // at_str == "approach"
+      action_type_ = APPROACH;
+    }
+
+    if (action_type_ == STOP) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".stop_points", rclcpp::ParameterValue(3));
+      stop_points_ = node->get_parameter(polygon_name_ + ".stop_points").as_int();
+    } else if (action_type_ == SLOWDOWN) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".stop_points", rclcpp::ParameterValue(3));
+      stop_points_ = node->get_parameter(polygon_name_ + ".stop_points").as_int();
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".slowdown", rclcpp::ParameterValue(0.5));
+      slowdown_ = node->get_parameter(polygon_name_ + ".slowdown").as_double();
```

Comment:
It is expressed in fraction (`0.35`) rather than in percentage (although, I consider it is the same `0.35 == 35%`). However, in order to avoid some confusion, I'd like to not using a percentage keyword. How about `slowdown_ratio` ?

Reply:
Sure

Reply:
[Done](https://github.com/ros-planning/navigation2/pull/2982/files#diff-6ff706b80d78c34ef10bd9efd8109834fe66be34241d395835af7c3857df760eR190-R192)